### PR TITLE
Cutpoints improved

### PR DESCRIPTION
**Pull request by Sh-Pe**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java [hunk 1]
```diff
@@ -67,4 +69,37 @@ public static double applyMagnitudeCompensation(Rotation2d velocityPerSecond, do
 			/ (magnitude + SwerveConstants.AIM_ASSIST_MAGNITUDE_FACTOR);
 	}
 
+	public static Translation2d getClosestShootingPoint(Translation2d robotPose, Pair<Rotation2d, Rotation2d>[] invalidRanges, double DistanceForShootingFromSpeaker) {
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java]: move it to file of its own (shooting service or something)

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java [hunk 2]
```diff
@@ -67,4 +69,37 @@ public static double applyMagnitudeCompensation(Rotation2d velocityPerSecond, do
...
+		Translation2d speaker = Field.getSpeaker().toTranslation2d();
+		Translation2d robotPoseRelativeToSpeaker = robotPose.minus(speaker);
+		Rotation2d angleFromSpeaker = Rotation2d.fromRadians(Math.atan2(robotPoseRelativeToSpeaker.getY(), robotPoseRelativeToSpeaker.getX()));
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java]: split to function

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java [hunk 3]
```diff
@@ -67,4 +69,37 @@ public static double applyMagnitudeCompensation(Rotation2d velocityPerSecond, do
...
+				}
+
+				closestValidPointRelativeToSpeaker = new Translation2d(Math.cos(closestValidAngle.getRadians()), Math.sin(closestValidAngle.getRadians())).times(DistanceForShootingFromSpeaker);
+				return closestValidPointRelativeToSpeaker.plus(speaker);
+			}
+		}
+
+		double slope = robotPoseRelativeToSpeaker.getY() / robotPoseRelativeToSpeaker.getX();
+		double closestValidPointX = DistanceForShootingFromSpeaker / Math.sqrt(Math.pow(slope, 2) + 1);
+		closestValidPointRelativeToSpeaker = new Translation2d(closestValidPointX, slope * closestValidPointX);
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java]: split to function

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java [hunk 4]
```diff
@@ -67,4 +69,37 @@ public static double applyMagnitudeCompensation(Rotation2d velocityPerSecond, do
...
+				Rotation2d closestValidAngle;
+				if (Math.abs(angleFromSpeakerRadians - minimumRangeRadians) <= (angleFromSpeakerRadians - maximumRangeRadians)) {
+					closestValidAngle = invalidRange.getFirst();
+				} else {
+					closestValidAngle = invalidRange.getSecond();
+				}
+
+				closestValidPointRelativeToSpeaker = new Translation2d(Math.cos(closestValidAngle.getRadians()), Math.sin(closestValidAngle.getRadians())).times(DistanceForShootingFromSpeaker);
+				return closestValidPointRelativeToSpeaker.plus(speaker);
+			}
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java]: split to function

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java (hunk 1 above)
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java]: add in the file constants of the invalid ranges, and the shooting radius
and add func thaat uses them and gets only robot

Comment (YHerm) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/AimAssistMath.java]:
```suggestion
	public static Translation2d getClosestShootingPoint(Translation2d robotPose, double DistanceForShootingFromSpeaker, Pair<Rotation2d, Rotation2d>... invalidRanges) {
```

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/ShootingHelpers/ShootingHelpersConstants.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package frc.robot.subsystems.swerve.swervestatehelpers.ShootingHelpers;
+
+public class ShootingHelpersConstants {
+
+	protected static final double SHOOTING_DISTANCE_FROM_SPEAKER = 0;
+
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/ShootingHelpers/ShootingHelpersConstants.java]: add the ranges (like 20 degrees for the sides for now)

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/ShootingHelpers/ShootingHelpers.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		return Rotation2d.fromRadians(Math.atan2(robotPoseRelativeToSpeaker.getY(), robotPoseRelativeToSpeaker.getX()));
+	}
+
+	public static Translation2d getCutPointOnRadiusFromCoordinates(Translation2d point, double radius) {
+		double slope = point.getY() / point.getX();
+		double closestValidPointX = radius / Math.sqrt(Math.pow(slope, 2) + 1);
+		return new Translation2d(closestValidPointX, slope * closestValidPointX);
+	}
+
+	public static Rotation2d getClosestAngleWithinRange(Rotation2d minimumRange, Rotation2d maximumRange, Rotation2d angleWithinRange) {
```
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/ShootingHelpers/ShootingHelpers.java]:
```suggestion
	public static Rotation2d getClosestAngleToRange(Rotation2d minimumRange, Rotation2d maximumRange, Rotation2d angleWithinRange) {
```

**Review by YHerm — COMMENTED** [review]
make sure to test it on both sides (red and blue) speaker

File: src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/ShootingHelpers/ShootingHelpers.java
Comment (YHerm) [inline src/main/java/frc/robot/subsystems/swerve/swervestatehelpers/ShootingHelpers/ShootingHelpers.java:1]: put package outside of swerve and subsystems
also packages names are alllowerletters

**Comment by Sh-Pe** [comment]
on need robot testing since its and deterministic algo, and I've checked the code on the simulation.

File: src/main/java/frc/utils/shootinghelpers/ShootingHelpers.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+			)
+		);
+	}
+
+	@SafeVarargs
+	public static Translation2d getClosestShootingPoint(
+		Translation2d robotPosition,
+		double DistanceForShootingFromSpeaker,
+		Translation2d speakerPosition,
+		Pair<Rotation2d, Rotation2d>... invalidRanges
```
Comment (Yotam7) [inline src/main/java/frc/utils/shootinghelpers/ShootingHelpers.java:38]: write javadoc that mentions that invaildRanges must not be overlapping

File: src/main/java/frc/utils/shootinghelpers/ShootingHelpers.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	) {
+		Translation2d speaker = speakerPosition;
```
Comment (Yotam7) [inline src/main/java/frc/utils/shootinghelpers/ShootingHelpers.java]: no need to rename variable
```suggestion
```

**Comment by Sh-Pe** [comment]
added those by mistake

**Review by drorta — DISMISSED** [review]
Test it if u didnt

**Comment by Sh-Pe** [comment]
> Test it if u didnt

I did test it with Desmos and souts

File: src/main/java/frc/utils/shootinghelpers/ShootingHelpers.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+		Translation2d robotPosition,
+		double DistanceForShootingFromSpeaker,
+		Translation2d speakerPosition,
+		Pair<Rotation2d, Rotation2d>... invalidRanges
+	) {
+		Translation2d robotRelativeToSpeaker = robotPosition.minus(speakerPosition);
+		Rotation2d angleFromSpeaker = findAngleFromSpeaker(robotPosition, speakerPosition);
+		double angleFromSpeakerRadians = angleFromSpeaker.getRadians();
+
+		Translation2d closestValidPointRelativeToSpeaker;
```
Comment (hillelv) [inline src/main/java/frc/utils/shootinghelpers/ShootingHelpers.java:44]: why do you declare it here

File: src/main/java/frc/utils/shootinghelpers/ShootingHelpers.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+		return Rotation2d.fromRadians(Math.atan2(robotPositionRelativeToSpeaker.getY(), robotPositionRelativeToSpeaker.getX()));
+	}
+
+	private static Translation2d getCutPointOnRadiusFromCoordinates(Translation2d point, double radius) {
+		double slope = point.getY() / point.getX();
+		double closestValidPointX = radius / Math.sqrt(Math.pow(slope, 2) + 1);
+		return new Translation2d(closestValidPointX, slope * closestValidPointX);
+	}
+
+	private static Rotation2d getClosestAngleIntervalBound(Rotation2d minimumRange, Rotation2d maximumRange, Rotation2d angleWithinRange) {
```
Comment (hillelv) [inline src/main/java/frc/utils/shootinghelpers/ShootingHelpers.java:75]: bool b = coplex boolean
return b ? min : max

**Comment by greenblitz4590** [comment]
hillel - why do you resolve without changing or explaining 😠 grrr

**Pull request closed without merging**